include/envoy/config/typed_metadata.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/stream_info/filter_state.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * TypedMetadata interface.
+ */
+class TypedMetadata {
+public:
+  virtual ~TypedMetadata() {}
+
+  // Returns a T instance by key. If the conversion is not able to complete, or
+  // if the data is not in the store, returns a nullptr.
+  template <typename T> const T* get(const std::string& key) const {
+    static_assert(std::is_base_of<StreamInfo::FilterState::Object, T>::value,
+                  "Data type must be subclass of StreamInfo::FilterState::Object");
+    const StreamInfo::FilterState::Object* p = getData(key);
+    if (p != nullptr) {
+      return dynamic_cast<const T*>(p);
+    }
+    return nullptr;
+  }
+
+protected:
+  // Returns data associated with given 'key'.
+  virtual const StreamInfo::FilterState::Object* getData(const std::string& key) const PURE;
+};
+
+/*
+ * Typed metadata should implement this factory and register via Registry::registerFactory or the
+ * convenience class RegisterFactory.
+ */
+class TypedMetadataFactory {
+public:
+  virtual ~TypedMetadataFactory() {}
+
+  // Name of the factory.
+  // It's used as key in the metadata map, as well as key in the factory registry.
+  // When building a TypedMetadata from envoy::api::v2::core::Metadata, if the key is not found, the
+  // parse will not be called and the corresponding typedMetadata entry will be set to nullptr.
+  virtual const std::string name() const PURE;
+
+  // Convert the google.protobuf.Struct into an instance of StreamInfo::FilterState::Object.
+  // It should throw an EnvoyException in case the conversion can't be completed.
+  // Returns a derived class object pointer of TypedMetadata.
+  virtual std::unique_ptr<const StreamInfo::FilterState::Object>
+  parse(const ProtobufWkt::Struct&) const PURE;
```

Comment:
Why this reliance on FilterState object? The whole idea is to be able to convert arbitrary metadata into typed variants in order to avoid unnecessary conversion in the data path. 
Also, how are you expecting end users/authors to create their proto messages that inherit from FilterStateObject? Or am I missing something here?

I had envisioned something along the lines of this:
 metadata entry is a Struct which wraps Any.
 You take the Any inside the struct and cast it into the proto Type, giving all the subsystems full access to the proto form of the config metadata. 

Take a look at how the per-route fault filter thing is done. It does a struct to proto conversion in the route. This is something that should be cleaned up with this PR. There are other subsystems doing a similar conversion, storing the object here and there.

Reply:
> Assuming this does a round trip through JSON, is the intent to call this once during config load, store it as part of route/cluster and then hand it over to filters as typed objects, in order to avoid the conversion at runtime?
> This is similar to what @qiwzhang was proposing (albeit a bit cleaner)
Not sure what's the other proposal, a link please?

Reply:
> Why this reliance on FilterState object? The whole idea is to be able to convert arbitrary metadata into typed variants in order to avoid unnecessary conversion in the data path.

It's just a code reusing. I can definitely define a new top level object of same definition, or possibly move the Object out of the file. 

> Also, how are you expecting end users/authors to create their proto messages that inherit from FilterStateObject? Or am I missing something here?

Yes. Data entries stored in typed_metadata_ store has to have a top level type. For now it's set to FilterState::Object.

> 
> I had envisioned something along the lines of this:
> metadata entry is a Struct which wraps Any.
> You take the Any inside the struct and cast it into the proto Type, giving all the subsystems full access to the proto form of the config metadata.
Yes.
> 
> Take a look at how the per-route fault filter thing is done. It does a struct to proto conversion in the route. This is something that should be cleaned up with this PR. There are other subsystems doing a similar conversion, storing the object here and there.
Could you give me some example? I did notice that in filterState::get we are doing similar conversion work, but cannot reuse the code as it throws. 

The idea here is that not necessarily all metadata entry has to be translated/de-serialized, only entries folks care about(register a metadataFactory for it) will be converted into typed data.

Reply:
> In in this code, its all metadata in route.

It's not. 
Currently there are RouteEntry and ClusterInfo that has the "core.Metadata metadata();" interface. I am adding typedMetadata() interface to these two classes in this PR. 

> If I have my custom filter that has additional metadata configured through the control plane, how would I use this feature?

In that case, say you want to have a helper class for your CLUSTER metadata with key "Baz", 
you could register a BazTypedMetadataFactory which extends from ClusterTypedMetadataFactory, like the example upstream_impl_test.cc does.
http://go/gh/envoyproxy/envoy/pull/4723/files#diff-5560832ea7741e9a4298f2f4d4dfdae5
When reloading the CDS config, the new config_impl instance will create the new wrapper instance and store it.

> Second, if I insert, typed protobufs into the route, such as per-route fault filter (which matches the protobuf fault filter type), how would I take advantage of this code? Is every filter author required to change all their code to take advantage of this type cast feature?
> 

I think Route.perFilterConfig(if that's what you mean) serves different purpose than route metadata? This PR is adding additional API to make better use of metadata. If a user don't require customized logic around the metadata, they don't necessarily have to provide the translation. (and the typedMetadata returns nullptr in this case).

> Take a look at the perFilterConfig stuff (https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L609).. Its not just in route. ITs in virtualhosts as well. None of that is being typecasted. I believe that this whole typed metadata (with reverse dns filter names being the key) is actually better fit in the per filter config [https://github.com/envoyproxy/envoy/blob/86355a667aa11130688fb9d91902a42b96c40dc2/include/envoy/router/router.h#L304] which right now is forcing filter authors to add another object type in their filters.
> So how does the current PR which is forcing filter authors to return StreamINfo::Object differ ? Why have two different mechanisms in the same code, for very similar/same purpose?

Again It's optional, not mandatory.